design/secure-communication.md

```diff
@@ -0,0 +1,94 @@
+
+It's also important to note that the certificates talked about in this document are **only** the certificates used to secure the gRPC channel between Contour and Envoy, not any certificates used to serve TLS out of Envoy. As such, this design is wholly separate to any certificate functionality on either Ingress or IngressRoute.
+
+## High-Level Design
+
+We will add configuration to Contour, both for `contour serve` and `contour bootstrap`, to have Contour and Envoy use a shared CA and a keypair each. This will allow a basic split deployment with minimal changes to the current code, using simple Secret mounts into file locations in both a Contour Deployment and Envoy deployment or DaemonSet.
+
+The generation of the certificates and their placement into Secrets in an operable and maintainable way is not in scope for this change, only a basic illustration in the example deployment YAMLs and Howto. However, see [Future Work](#future-work) for more info.
+
+## Detailed Design - mTLS
+
+In order for mTLS to work, each end must know:
+- The CA that the TLS relationship will descend from.
+- Its certificate
+- Its key
+
+This applies both to Contour and Envoy.
+
+To accomplish this, we will add the following new command line options:
+
+To `contour serve`, covering the Contour side:
+- `--contour-cafile`: the CA Bundle for the mutual TLS relationship
+- `--contour-mtls-cert`: The cert for Contour to use to identify itself. (`serve`)
+- `--contour-mtls-key`: The key for Contour to use for its side of the mTLS pairing. (`serve`)
+
+And to `contour bootstrap`, covering Envoy configuration:
+- `--envoy-cafile`: The CA Bundle for the mutual TLS relationship
+- `--envoy-mtls-cert`: The cert for Envoy to use to identify itself.
+- `--envoy-mtls-key`: The key for Envoy to use to identify itself.
+
+If any one of the three options is passed to either `serve` or `bootstrap`, the other two are required, and we will error out if they are not passed.
+
+We will wire these options into the gRPC server establishment process on the Contour side, and supply them in the bootstrap on the Envoy side. This is where the most significant code changes will be required.
+
+We will update the example deployment manifests for the separate deployment options to use these new options, and provide a basic sample howto on generating a CA keypair and mTLS keypairs using command line tools like openssl.
+
+## Security Considerations
+
+In the current design, these certs require the rolling of the deployment to change, they should be long-lived certs (days, weeks, or months), not short-lived (minutes or hours). The CA Keypair should be very long-lived and very tightly controlled, as Contour/Envoy connection security is only as secure as the CA keypair.
```

Comment:
s/these certs require the rolling of the deployment to change/the process must be restarted to pick a change to these certs/